bindings/rust/s2n-tls/src/callbacks/async_cb.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Support for asynchronous callbacks.
+//!
+//! The general flow for an async callback is:
+//! 1. The application sets FooCallback on the [`crate::config::Config`] with
+//!    a method like Config::set_foo_callback.
+//! 2. When the underlying C library reaches the trigger for that specific
+//!    callback (for example, the ClientHello for [`crate::callbacks::ClientHelloCallback`])
+//!    it calls the callback implementation to get a [`ConnectionFuture`].
+//! 3. The [`ConnectionFuture`] is stored on the connection. Every time
+//!    the handshake is polled, the [`ConnectionFuture`] is polled instead.
+//! 4. Once the [`ConnectionFuture`] returns a result, the connection
+//!    drops the future and proceeds as usual.
+
+use crate::{connection::Connection, enums::CallbackResult, error::Error};
+use core::task::Poll;
+use pin_project_lite::pin_project;
+use std::pin::Pin;
+
+/// The Future associated with the async connection callback.
+///
+/// The calling application can provide an instance of [`ConnectionFuture`]
+/// when implementing an async callback, eg. [`crate::callbacks::ClientHelloCallback`],
+/// if it wants to run an asynchronous operation (disk read, network call).
+/// The application can return an error ([Err(Error::application())])
+/// to indicate connection failure.
+pub trait ConnectionFuture {
+    fn poll(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>>;
+}
+
+pub(crate) type ConnectionFutureResult = Result<Option<Pin<Box<dyn ConnectionFuture>>>, Error>;
+
+// Useful for propagating [`error::Error`] from a C callback back to the Rust application.
+pub(crate) struct ErrorFuture {
+    error: Option<Error>,
+}
+
+impl ConnectionFuture for ErrorFuture {
+    fn poll(
+        mut self: Pin<&mut Self>,
+        _connection: &mut Connection,
+        _ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>> {
+        let err = self.error.take().expect(
+            "ErrorFuture should be initialized with Some(error) and a Future should never
+            be polled after it returns Poll::Ready",
+        );
+        Poll::Ready(Err(err))
+    }
+}
+
+pin_project! {
+    /// A wrapper around an optional [`ConnectionFuture`]
+    /// which either polls the future or immediately reports success.
+    struct OptionalFuture{
+        #[pin]
+        option: Option<Pin<Box<dyn ConnectionFuture>>>,
```

Comment:
Pinning an Option\<Pin\> feels like it might be overkill.

Reply:
Why does it need to be pinned?

Reply:
I'm rereading https://doc.rust-lang.org/nightly/core/pin/index.html#projections-and-structural-pinning to try to answer that question, and I *think* the answer is "it doesn't" because [Optional is Unpin](https://doc.rust-lang.org/std/option/enum.Option.html#impl-Unpin-for-Option%3CT%3E)?